clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerHeartbeatRequestManager.java

```diff
@@ -211,6 +214,29 @@ public ConsumerMembershipManager membershipManager() {
         return membershipManager;
     }
 
+    @Override
+    protected boolean shouldSendLeaveHeartbeatNow() {
+        // If the consumer has dynamic membership,
+        // we should skip the leaving heartbeat when leaveGroupOperation is REMAIN_IN_GROUP
+        if (membershipManager.groupInstanceId().isEmpty() && REMAIN_IN_GROUP == membershipManager.leaveGroupOperation())
+            return false;
+        return membershipManager().state() == MemberState.LEAVING;
+    }
+
+    @Override
+    public NetworkClientDelegate.PollResult pollOnClose(long currentTimeMs) {
+        // Determine if we should send a leaving heartbeat:
+        // - For static membership (when groupInstanceId is present): Always send the leaving heartbeat
+        // - For dynamic membership: Send the leaving heartbeat only when leaveGroupOperation is not REMAIN_IN_GROUP
+        boolean shouldHeartbeat = membershipManager.groupInstanceId().isPresent()
```

Comment:
IIUC the only difference in this #pollOnClose implementation from the original one in the abstract class is the addition of this `shouldHeartbeat` condition to the `if membershipManager().isLeavingGroup()` check. 

However it seems like the extra condition is built entirely of calls to `membershipManager` methods so I'm wondering if it doesn't make more sense to just incorporate the `shouldHeartbeat` condition into the `#isLeavingGroup` method of the ConsumerMembershipManager class? That way we don't need to make `#pollOnClose` abstract and can cut down on some redundant code

Reply:
Also from a quick look at the ConsumerMembershipManager#isLeavingGroup` implementation, it already includes checks on both of these conditions (`groupInstanceId.isPresent` and `REMAIN_IN_GROUP == membershipManager.leaveGroupOperation()`). imo it would be cleaner and easier to read if we just encapsulated all the logic around these conditions inside the `#isLeavingGroup` method

Reply:
Yes, this could significantly reduce redundancy!
Previously, I thought it was more clear that the async consumer and share consumer have their implementation, but it does not provide too much value now.